File: apps/reading/src/api/workstationRecordPromptModelDecisionCompose.ts
/**
 * Workstation record → prompt context → model decision pack (pure).
 *
 * Operator vision: while wrestling in the research workstation, recorded
 * insights/questions recursively inform prompts; operator chooses model with
 * usage bar + projection of how the enriched prompt affects budget.
 *
 * record_persisted always false.
 * prompts_injected always false.
 * live_router_authorized always false.
 * secrets_stored always false.
 * live_meter_read always false.
 */

import {
  composeWorkstationSessionInsightRecord,
  type SessionRecordItem,
  type WorkstationSessionInsightRecordCompose,
} from "./workstationSessionInsightRecordCompose";
import {
  bridgeWorkstationRecordPromptContext,
  type ContextPlacement,
  type PromptContextEnvelope,
} from "./workstationRecordPromptContextBridge";
import type { WorkstationRecordItem } from "./workstationRecursiveRecordPack";
import {
  composeSettingsDecisionTreeUsageBar,
  type SettingsDecisionTreeUsageBarCompose,
} from "./settingsDecisionTreeUsageBarCompose";
import type { ModelOption } from "./modelDecisionPromptCompose";
import type {
  BenchTaskBest,
  NotDiamondShadowRec,
} from "./settingsModelDriverTabCompose";

export interface WorkstationRecordPromptModelDecisionInput {
  session_id: string;
  parent_asset_id: string;
  records: SessionRecordItem[];
  user_prompt: string;
  placement?: ContextPlacement;
  max_context_lines?: number | null;
  selected_model_id: string;
  models: ModelOption[];
  daily_cap_usd: number | null;
  spent_usd: number | null;
  /**
   * Optional projected cost for the *enriched* prompt. When null, projection
   * honesty stays null (never invent $0).
   */
  projected_cost_usd_high?: number | null;
  projected_cost_usd_low?: number | null;
  bench_bests?: BenchTaskBest[] | null;
  focus_task?: string | null;
  nd_shadow?: NotDiamondShadowRec | null;
  operator_ack: boolean;
}

export interface WorkstationRecordPromptModelDecisionCompose {
  session_id: string;
  parent_asset_id: string;
  records: WorkstationSessionInsightRecordCompose;
  bridge: PromptContextEnvelope;
  decision: SettingsDecisionTreeUsageBarCompose;
  /**
   * True when records.record_ready, bridge.bridge_ready, decision.decision_ready.
   * Still never injects prompts or routes live.
   */
  pack_ready: boolean;
  proposed_prompt: string;
  would_exceed: boolean | null;
  usage_percent: number | null;
  record_persisted: false;
  prompts_injected: false;
  live_router_authorized: false;
  secrets_stored: false;
  live_meter_read: false;
  notes: string[];
  authority: "workstation_record_prompt_model_decision_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function mapSessionToPackItems(
  records: SessionRecordItem[],
): WorkstationRecordItem[] {
  return records.map((r) => {
    let kind: WorkstationRecordItem["kind"];
    switch (r.kind) {
      case "insight":
        kind = "insight";
        break;
      case "question":
        kind = "question";
        break;
      case "data":
      case "claim":
        kind = "finding";
        break;
      default:
        throw new Error(`unsupported session record kind: ${String(r.kind)}`);
    }
    const item: WorkstationRecordItem = {
      record_id: r.record_id,
      kind,
      text: r.body,
    };
    if (r.source_ref) {
      item.asset_id = r.source_ref;
    }
    return item;
  });
}

/**
 * Compose session records → prompt context envelope → model decision usage bar.
 * Never persists, injects, or live-routes.
 */
export function composeWorkstationRecordPromptModelDecision(
  input: WorkstationRecordPromptModelDecisionInput,
): WorkstationRecordPromptModelDecisionCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );

  const notes: string[] = [
    "record_persisted=false — records are pure pack only",
    "prompts_injected=false — proposed envelope only",
    "live_router_authorized=false — operator selects model",
    "secrets_stored=false",
    "live_meter_read=false",
  ];

  const records = composeWorkstationSessionInsightRecord({
    session_id,
    parent_asset_id,
    records: input.records,
    operator_ack: input.operator_ack,
    mark_for_prompt_context: true,
  });
  notes.push(...records.notes);

  const packItems = mapSessionToPackItems(input.records);
  const bridge = bridgeWorkstationRecordPromptContext({
    session_id,
    user_prompt: input.user_prompt,
    items: packItems,
    max_context_lines: input.max_context_lines,
    placement: input.placement,
    model_decision: {
      selected_model_id: input.selected_model_id,
      models: input.models,
      daily_cap_usd: input.daily_cap_usd,
      spent_usd: input.spent_usd,
      projected_cost_usd_high: input.projected_cost_usd_high,
      projected_cost_usd_low: input.projected_cost_usd_low,
    },
  });
  notes.push(...bridge.notes);

  const decision = composeSettingsDecisionTreeUsageBar({
    selected_model_id: input.selected_model_id,
    models: input.models,
    daily_cap_usd: input.daily_cap_usd,
    spent_usd: input.spent_usd,
    projected_cost_usd_high: input.projected_cost_usd_high,
    projected_cost_usd_low: input.projected_cost_usd_low,
    bench_bests: input.bench_bests,
    focus_task: input.focus_task,
    nd_shadow: input.nd_shadow,
    operator_ack: input.operator_ack,
  });
  notes.push(...decision.notes);

  // Prefer would_exceed from decision bar (same as bridge model_decision if set)
  const would_exceed =
    decision.would_exceed ??
    (bridge.model_decision ? bridge.model_decision.would_exceed : null);

  const pack_ready =
    records.record_ready &&
    bridge.bridge_ready &&
    decision.decision_ready &&
    records.record_persisted === false &&
    bridge.prompts_injected === false &&
    decision.live_router_authorized === false;

  if (!records.record_ready) {
    notes.push("pack_ready=false — session records not ready");
  } else if (!bridge.bridge_ready) {
    notes.push("pack_ready=false — prompt context bridge not ready");
  } else if (!decision.decision_ready) {
    notes.push("pack_ready=false — model decision tree not ready");
  } else {
    notes.push(
      "pack_ready=true — records→prompt→model intent only; still pure",
    );
  }

  if (
    records.record_persisted !== false ||
    records.prompts_injected !== false ||
    bridge.prompts_injected !== false ||
    bridge.record_persisted !== false ||
    decision.live_router_authorized !== false ||
    decision.secrets_stored !== false ||
    decision.live_meter_read !== false
  ) {
    throw new Error("invariant: nested honesty flags must remain false");
  }

  notes.push("record_persisted=false");
  notes.push("prompts_injected=false");
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");

  return {
    session_id,
    parent_asset_id,
    records,
    bridge,
    decision,
    pack_ready,
    proposed_prompt: bridge.proposed_prompt,
    would_exceed,
    usage_percent: decision.usage_percent,
    record_persisted: false,
    prompts_injected: false,
    live_router_authorized: false,
    secrets_stored: false,
    live_meter_read: false,
    notes,
    authority: "workstation_record_prompt_model_decision_compose_advisory",
  };
}

export function formatWorkstationRecordPromptModelDecisionSummary(
  c: WorkstationRecordPromptModelDecisionCompose,
): string {
  const w =
    c.would_exceed === null
      ? "would_exceed=null"
      : `would_exceed=${c.would_exceed}`;
  const pct =
    c.usage_percent === null
      ? "usage%=null"
      : `usage%=${c.usage_percent.toFixed(1)}`;
  return (
    `pack_ready=${c.pack_ready} · records=${c.records.record_count} · ` +
    `model=${c.decision.driver.decision.selected_model_id} · ${pct} · ${w} · ` +
    `record_persisted=false · prompts_injected=false · live_router_authorized=false · ` +
    `secrets_stored=false · live_meter_read=false`
  );
}

File: substrate/workstation_record_prompt_model_decision_compose.py
"""Workstation record → prompt context → model decision pack (pure).

record_persisted, prompts_injected, live_router_authorized, secrets_stored,
live_meter_read always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.settings_decision_tree_usage_bar_compose import (
    SettingsDecisionTreeUsageBarCompose,
    SettingsDecisionTreeUsageBarComposeError,
    compose_settings_decision_tree_usage_bar,
)
from substrate.workstation_record_prompt_context_bridge import (
    PromptContextEnvelope,
    WorkstationRecordPromptContextBridgeError,
    bridge_workstation_record_prompt_context,
)
from substrate.workstation_session_insight_record_compose import (
    WorkstationSessionInsightRecordCompose,
    WorkstationSessionInsightRecordComposeError,
    compose_workstation_session_insight_record,
)

KIND_MAP = {
    "insight": "insight",
    "question": "question",
    "data": "finding",
    "claim": "finding",
}


class WorkstationRecordPromptModelDecisionComposeError(ValueError):
    """Fail-closed validation for record→prompt→model pack."""


@dataclass(frozen=True)
class WorkstationRecordPromptModelDecisionCompose:
    session_id: str
    parent_asset_id: str
    records: WorkstationSessionInsightRecordCompose
    bridge: PromptContextEnvelope
    decision: SettingsDecisionTreeUsageBarCompose
    pack_ready: bool
    proposed_prompt: str
    would_exceed: bool | None
    usage_percent: float | None
    record_persisted: bool
    prompts_injected: bool
    live_router_authorized: bool
    secrets_stored: bool
    live_meter_read: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "records": self.records.to_dict(),
            "bridge": self.bridge.to_dict(),
            "decision": self.decision.to_dict(),
            "pack_ready": self.pack_ready,
            "proposed_prompt": self.proposed_prompt,
            "would_exceed": self.would_exceed,
            "usage_percent": self.usage_percent,
            "record_persisted": False,
            "prompts_injected": False,
            "live_router_authorized": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "notes": list(self.notes),
            "authority": "workstation_record_prompt_model_decision_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise WorkstationRecordPromptModelDecisionComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _map_session_to_pack_items(records: list[object]) -> list[dict[str, Any]]:
    items: list[dict[str, Any]] = []
    for i, r in enumerate(records):
        if not isinstance(r, dict):
            raise WorkstationRecordPromptModelDecisionComposeError(
                f"records[{i}] must be an object"
            )
        kind = r.get("kind")
        if kind not in KIND_MAP:
            raise WorkstationRecordPromptModelDecisionComposeError(
                f"unsupported session record kind: {kind!r}"
            )
        item: dict[str, Any] = {
            "record_id": r.get("record_id"),
            "kind": KIND_MAP[str(kind)],
            "text": r.get("body"),
        }
        if r.get("source_ref"):
            item["asset_id"] = r.get("source_ref")
        items.append(item)
    return items


def compose_workstation_record_prompt_model_decision(
    *,
    session_id: object,
    parent_asset_id: object,
    records: object,
    user_prompt: object,
    selected_model_id: object,
    models: object,
    daily_cap_usd: object,
    spent_usd: object,
    operator_ack: object,
    placement: object | None = None,
    max_context_lines: object | None = None,
    projected_cost_usd_high: object | None = None,
    projected_cost_usd_low: object | None = None,
    bench_bests: object | None = None,
    focus_task: object | None = None,
    nd_shadow: object | None = None,
) -> WorkstationRecordPromptModelDecisionCompose:
    """Compose records → prompt envelope → model decision. Never injects/routes."""
    if not isinstance(operator_ack, bool):
        raise WorkstationRecordPromptModelDecisionComposeError(
            "operator_ack must be an explicit boolean"
        )
    session = _require_nonempty(session_id, field="session_id")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")
    if not isinstance(records, list):
        raise WorkstationRecordPromptModelDecisionComposeError(
            "records must be an array"
        )

    notes: list[str] = [
        "record_persisted=false — records are pure pack only",
        "prompts_injected=false — proposed envelope only",
        "live_router_authorized=false — operator selects model",
        "secrets_stored=false",
        "live_meter_read=false",
    ]

    try:
        record_pack = compose_workstation_session_insight_record(
            session_id=session,
            parent_asset_id=parent,
            records=records,
            operator_ack=operator_ack,
            mark_for_prompt_context=True,
        )
    except WorkstationSessionInsightRecordComposeError as e:
        raise WorkstationRecordPromptModelDecisionComposeError(str(e)) from e
    notes.extend(record_pack.notes)

    pack_items = _map_session_to_pack_items(records)
    try:
        bridge = bridge_workstation_record_prompt_context(
            session_id=session,
            user_prompt=user_prompt,
            items=pack_items,
            max_context_lines=max_context_lines,
            placement=placement,
            model_decision={
                "selected_model_id": selected_model_id,
                "models": models,
                "daily_cap_usd": daily_cap_usd,
                "spent_usd": spent_usd,
                "projected_cost_usd_high": projected_cost_usd_high,
                "projected_cost_usd_low": projected_cost_usd_low,
            },
        )
    except WorkstationRecordPromptContextBridgeError as e:
        raise WorkstationRecordPromptModelDecisionComposeError(str(e)) from e
    notes.extend(bridge.notes)

    try:
        decision = compose_settings_decision_tree_usage_bar(
            selected_model_id=selected_model_id,
            models=models,
            daily_cap_usd=daily_cap_usd,
            spent_usd=spent_usd,
            projected_cost_usd_high=projected_cost_usd_high,
            projected_cost_usd_low=projected_cost_usd_low,
            bench_bests=bench_bests,
            focus_task=focus_task,
            nd_shadow=nd_shadow,
            operator_ack=operator_ack,
        )
    except SettingsDecisionTreeUsageBarComposeError as e:
        raise WorkstationRecordPromptModelDecisionComposeError(str(e)) from e
    notes.extend(decision.notes)

    would_exceed = decision.would_exceed
    if would_exceed is None and bridge.model_decision is not None:
        would_exceed = bridge.model_decision.would_exceed

    pack_ready = (
        record_pack.record_ready
        and bridge.bridge_ready
        and decision.decision_ready
        and record_pack.record_persisted is False
        and bridge.prompts_injected is False
        and decision.live_router_authorized is False
    )
    if not record_pack.record_ready:
        notes.append("pack_ready=false — session records not ready")
    elif not bridge.bridge_ready:
        notes.append("pack_ready=false — prompt context bridge not ready")
    elif not decision.decision_ready:
        notes.append("pack_ready=false — model decision tree not ready")
    else:
        notes.append(
            "pack_ready=true — records→prompt→model intent only; still pure"
        )

    if (
        record_pack.record_persisted is not False
        or record_pack.prompts_injected is not False
        or bridge.prompts_injected is not False
        or bridge.record_persisted is not False
        or decision.live_router_authorized is not False
        or decision.secrets_stored is not False
        or decision.live_meter_read is not False
    ):
        raise WorkstationRecordPromptModelDecisionComposeError(
            "invariant: nested honesty flags must remain false"
        )

    notes.extend(
        (
            "record_persisted=false",
            "prompts_injected=false",
            "live_router_authorized=false",
            "secrets_stored=false",
            "live_meter_read=false",
        )
    )

    return WorkstationRecordPromptModelDecisionCompose(
        session_id=session,
        parent_asset_id=parent,
        records=record_pack,
        bridge=bridge,
        decision=decision,
        pack_ready=pack_ready,
        proposed_prompt=bridge.proposed_prompt,
        would_exceed=would_exceed,
        usage_percent=decision.usage_percent,
        record_persisted=False,
        prompts_injected=False,
        live_router_authorized=False,
        secrets_stored=False,
        live_meter_read=False,
        notes=tuple(notes),
        authority="workstation_record_prompt_model_decision_compose_advisory",
    )


def format_workstation_record_prompt_model_decision_summary(
    c: WorkstationRecordPromptModelDecisionCompose,
) -> str:
    w = (
        "would_exceed=null"
        if c.would_exceed is None
        else f"would_exceed={c.would_exceed}"
    )
    pct = (
        "usage%=null"
        if c.usage_percent is None
        else f"usage%={c.usage_percent:.1f}"
    )
    return (
        f"pack_ready={c.pack_ready} · records={c.records.record_count} · "
        f"model={c.decision.driver.decision.selected_model_id} · "
        f"{pct} · {w} · "
        f"record_persisted=false · prompts_injected=false · "
        f"live_router_authorized=false · secrets_stored=false · live_meter_read=false"
    )


__all__ = [
    "WorkstationRecordPromptModelDecisionCompose",
    "WorkstationRecordPromptModelDecisionComposeError",
    "compose_workstation_record_prompt_model_decision",
    "format_workstation_record_prompt_model_decision_summary",
]
